sympy/solvers/tests/test_ode.py

```diff
@@ -2934,8 +2936,207 @@ def test_issue_14395():
     assert dsolve(Derivative(f(x), x, x) + 9*f(x) - sec(x), f(x)) == sol
 
 
+def test_matrix_exp():
+    from sympy.matrices.dense import Matrix, eye, zeros
+    from sympy.solvers.ode import matrix_exp
+    t = Symbol('t')
+
+    for n in range(1, 6+1):
+        assert matrix_exp(zeros(n), t) == eye(n)
+
+    for n in range(1, 6+1):
+        A = eye(n)
+        expAt = exp(t) * eye(n)
+        assert matrix_exp(A, t) == expAt
+
+    for n in range(1, 6+1):
+        A = Matrix(n, n, lambda i,j: i+1 if i==j else 0)
+        expAt = Matrix(n, n, lambda i,j: exp((i+1)*t) if i==j else 0)
+        assert matrix_exp(A, t) == expAt
+
+    A = Matrix([[0, 1], [-1, 0]])
+    expAt = Matrix([[cos(t), sin(t)], [-sin(t), cos(t)]])
+    assert matrix_exp(A, t) == expAt
+
+    A = Matrix([[2, -5], [2, -4]])
+    expAt = Matrix([
+            [3*exp(-t)*sin(t) + exp(-t)*cos(t), -5*exp(-t)*sin(t)],
+            [2*exp(-t)*sin(t), -3*exp(-t)*sin(t) + exp(-t)*cos(t)]
+            ])
+    assert matrix_exp(A, t) == expAt
+
+    A = Matrix([[21, 17, 6], [-5, -1, -6], [4, 4, 16]])
+    expAt = Matrix([
```

Comment:
This ```expAt``` isn't used. Should I remove or update this?

Reply:
I guess there should be an assert for this

Reply:
If you look below many things are commented. Those are things that should work and my intention was to fix all those by the time this PR got merged.

Reply:
Then what would be the ```A``` for this particular ```expAt```?

Reply:
Presumably the one right above

Reply:
It is not the case actually. The one just below corresponds with the A mentioned in the code.

Reply:
We can work backwards using log:
```
In [70]:     expAt = Matrix([ 
    ...:         [(8*t*exp(12*t) + 5*exp(12*t) - 1)*exp(4*t)/4, 
    ...:          (8*t*exp(12*t) + 5*exp(12*t) - 5)*exp(4*t)/4, 
    ...:          (exp(12*t) - 1)*exp(4*t)/2], 
    ...:         [(-8*t*exp(12*t) - exp(12*t) + 1)*exp(4*t)/4, 
    ...:          (-8*t*exp(12*t) - exp(12*t) + 5)*exp(4*t)/4, 
    ...:          (-exp(12*t) + 1)*exp(4*t)/2], 
    ...:         [4*t*exp(16*t), 4*t*exp(16*t), exp(16*t)]])                                                                                   

In [71]: expAt.log()                                                                                                                           
Out[71]: 
⎡         ⎛ 4⋅t⎞        ⎛ 16⋅t⎞             ⎛ 4⋅t⎞        ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞⎤
⎢      log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠        5⋅log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠⎥
⎢2⋅t - ───────── + ────────────  2⋅t - ─────────── + ────────────  - ───────── + ──────────⎥
⎢          4            4                   4             4              2           2     ⎥
⎢                                                                                          ⎥
⎢          ⎛ 4⋅t⎞      ⎛ 16⋅t⎞               ⎛ 4⋅t⎞      ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞ ⎥
⎢       log⎝ℯ   ⎠   log⎝ℯ    ⎠          5⋅log⎝ℯ   ⎠   log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠ ⎥
⎢-2⋅t + ───────── - ──────────   -2⋅t + ─────────── - ──────────    ───────── - ────────── ⎥
⎢           4           4                    4            4             2           2      ⎥
⎢                                                                                          ⎥
⎢                                                                            ⎛ 16⋅t⎞       ⎥
⎣             4⋅t                              4⋅t                        log⎝ℯ    ⎠       ⎦

In [72]: tp = Symbol('t', positive=True)                                                                                                       

In [73]: expAt.log().subs(t, tp)/tp                                                                                                            
Out[73]: 
⎡21  17  6 ⎤
⎢          ⎥
⎢-5  -1  -6⎥
⎢          ⎥
⎣4   4   16⎦
```

Reply:
We can work backwards using log:
```
In [70]:     expAt = Matrix([ 
    ...:         [(8*t*exp(12*t) + 5*exp(12*t) - 1)*exp(4*t)/4, 
    ...:          (8*t*exp(12*t) + 5*exp(12*t) - 5)*exp(4*t)/4, 
    ...:          (exp(12*t) - 1)*exp(4*t)/2], 
    ...:         [(-8*t*exp(12*t) - exp(12*t) + 1)*exp(4*t)/4, 
    ...:          (-8*t*exp(12*t) - exp(12*t) + 5)*exp(4*t)/4, 
    ...:          (-exp(12*t) + 1)*exp(4*t)/2], 
    ...:         [4*t*exp(16*t), 4*t*exp(16*t), exp(16*t)]])                                                                                   

In [71]: expAt.log()                                                                                                                           
Out[71]: 
⎡         ⎛ 4⋅t⎞        ⎛ 16⋅t⎞             ⎛ 4⋅t⎞        ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞⎤
⎢      log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠        5⋅log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠⎥
⎢2⋅t - ───────── + ────────────  2⋅t - ─────────── + ────────────  - ───────── + ──────────⎥
⎢          4            4                   4             4              2           2     ⎥
⎢                                                                                          ⎥
⎢          ⎛ 4⋅t⎞      ⎛ 16⋅t⎞               ⎛ 4⋅t⎞      ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞ ⎥
⎢       log⎝ℯ   ⎠   log⎝ℯ    ⎠          5⋅log⎝ℯ   ⎠   log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠ ⎥
⎢-2⋅t + ───────── - ──────────   -2⋅t + ─────────── - ──────────    ───────── - ────────── ⎥
⎢           4           4                    4            4             2           2      ⎥
⎢                                                                                          ⎥
⎢                                                                            ⎛ 16⋅t⎞       ⎥
⎣             4⋅t                              4⋅t                        log⎝ℯ    ⎠       ⎦

In [72]: tp = Symbol('t', positive=True)                                                                                                       

In [73]: expAt.log().subs(t, tp)/tp                                                                                                            
Out[73]: 
⎡21  17  6 ⎤
⎢          ⎥
⎢-5  -1  -6⎥
⎢          ⎥
⎣4   4   16⎦
```